# Session 10: Vector Search 101 — From Text to Numbers

## Module 4: RAG & Multimodal Systems

Learn how to transform text into numerical vectors, find similar documents instantly, and build a semantic search engine.

**Duration:** 3 hours | **Level:** Beginner

In [1]:
# Install required packages
# !pip install langchain-core langchain-google-genai faiss-cpu numpy -q

import os
import numpy as np
from langchain_google_genai import GoogleGenerativeAIEmbeddings

# Set your Google API key
os.environ['GOOGLE_API_KEY'] = 'AIzaSyCQMoesl69WYTyyyQteTAfTOwvFmc6fzdc'

# Initialize the embeddings model
embeddings = GoogleGenerativeAIEmbeddings(model='models/gemini-embedding-001')
print('✓ Embeddings model loaded successfully!')

✓ Embeddings model loaded successfully!


## What Are Embeddings?

An **embedding** is a numerical representation of text. Instead of working with raw words, machine learning models work with vectors (lists of numbers).

- **Text:** "I love machine learning"
- **Embedding:** [0.12, -0.45, 0.89, 0.23, ..., -0.34]  ← 768 numbers for Gemini

### Why Embeddings?
- **Semantic understanding:** Similar text → similar vectors
- **Distance-based search:** Find closest neighbors instantly
- **Works with any modality:** Text, images, audio (when trained appropriately)

### Key Insight
Embeddings capture *meaning*, not just syntax. "The cat sat on the mat" and "A feline rested on a carpet" will have similar embeddings!

In [2]:
# Create your first embedding
text = 'Hello, world!'
vector = embeddings.embed_query(text)

print(f'Text: {text}')
print(f'Embedding shape: {len(vector)} dimensions')
print(f'First 10 values: {vector[:10]}')
print(f'Vector dtype: {type(vector[0])}')

Text: Hello, world!
Embedding shape: 3072 dimensions
First 10 values: [-0.020920176, 0.009755219, 0.004780903, -0.059421252, 0.0050828396, 0.000864511, -0.0045847003, 0.0050770855, 0.019296365, 0.01731793]
Vector dtype: <class 'float'>


## Embedding Multiple Documents

Now let's embed a collection of related texts and see how their embeddings compare.

In [3]:
# Sample documents (travel-related)
documents = [
    'I love visiting tropical beaches in summer.',
    'Mountain hiking is my favorite outdoor activity.',
    'A sunny beach vacation is perfect for relaxation.',
    'Rock climbing is an exciting sport.',
    'Swimming in the ocean feels refreshing.',
    'Winter skiing in the Alps is incredible.'
]

# Embed all documents
doc_vectors = [embeddings.embed_query(doc) for doc in documents]

# Check the shape
print(f'Number of documents: {len(doc_vectors)}')
print(f'Dimensions per embedding: {len(doc_vectors[0])}')
print(f'Total size: {len(doc_vectors)} × {len(doc_vectors[0])} matrix')
print()
# Display first two documents with their embedding stats
for i, (doc, vec) in enumerate(zip(documents[:2], doc_vectors[:2])):
    print(f'Doc {i}: "{doc}"')
    print(f'  Mean: {np.mean(vec):.4f}, Std: {np.std(vec):.4f}')

Number of documents: 6
Dimensions per embedding: 3072
Total size: 6 × 3072 matrix

Doc 0: "I love visiting tropical beaches in summer."
  Mean: -0.0003, Std: 0.0180
Doc 1: "Mountain hiking is my favorite outdoor activity."
  Mean: 0.0001, Std: 0.0180


## Key Insight: Similar Text Produces Similar Vectors

Documents that are semantically similar should have embeddings that are close to each other in the vector space.

Notice in our sample:
- Documents about beaches (0, 2, 4) should cluster together
- Documents about mountains/snow (1, 3, 5) should cluster together

Let's verify this with similarity metrics!

## Similarity Metrics: Cosine Similarity

**Cosine similarity** measures the angle between two vectors, ranging from -1 to 1:
- **1.0** = identical direction (perfect similarity)
- **0.0** = perpendicular (no relation)
- **-1.0** = opposite direction

### Formula
```
cosine_sim(a, b) = (a · b) / (||a|| × ||b||)
```

### Why Cosine for Text?
- **Length-invariant:** "good" vs "really really good" are similar
- **Focuses on direction:** What matters is *which* dimensions are active
- **Industry standard:** Used in all modern semantic search systems

In [4]:
def cosine_similarity(a, b):
    """Compute cosine similarity between two vectors."""
    return np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b))

# Test with our embedded documents
print('Cosine Similarity Results')
print('=' * 60)

# Compare doc 0 (beach) with all others
query_doc = 0
similarities = []
for i, doc_vec in enumerate(doc_vectors):
    sim = cosine_similarity(doc_vectors[query_doc], doc_vec)
    similarities.append((i, documents[i], sim))
    print(f'{i}: {documents[i]}')
    print(f'   Similarity: {sim:.4f}')
    print()

# Sort by similarity
sorted_sims = sorted(similarities, key=lambda x: x[2], reverse=True)
print('Ranked by similarity to Doc 0:')
for rank, (idx, doc, sim) in enumerate(sorted_sims[:4]):
    print(f'  {rank+1}. Doc {idx}: {sim:.4f}')

Cosine Similarity Results
0: I love visiting tropical beaches in summer.
   Similarity: 1.0000

1: Mountain hiking is my favorite outdoor activity.
   Similarity: 0.6921

2: A sunny beach vacation is perfect for relaxation.
   Similarity: 0.7155

3: Rock climbing is an exciting sport.
   Similarity: 0.6244

4: Swimming in the ocean feels refreshing.
   Similarity: 0.6869

5: Winter skiing in the Alps is incredible.
   Similarity: 0.6190

Ranked by similarity to Doc 0:
  1. Doc 0: 1.0000
  2. Doc 2: 0.7155
  3. Doc 1: 0.6921
  4. Doc 4: 0.6869


## Euclidean Distance

**Euclidean distance** is the straight-line distance between points:

```
euclidean(a, b) = √(Σ(aᵢ - bᵢ)²)
```

### Comparison: Cosine vs Euclidean
- **Cosine:** Angle between vectors (ignores magnitude)
- **Euclidean:** Actual distance (considers magnitude)

For normalized embeddings, both metrics produce similar rankings!

In [5]:
def euclidean_distance(a, b):
    """Compute Euclidean distance between two vectors."""
    return np.linalg.norm(np.array(a) - np.array(b))

# Compare both metrics
print('Cosine Similarity vs Euclidean Distance')
print('=' * 70)

query_idx = 0
query_vec = doc_vectors[query_idx]
print(f'Query: \"{documents[query_idx]}\"')
print()

for i in range(len(documents)):
    cos_sim = cosine_similarity(query_vec, doc_vectors[i])
    euc_dist = euclidean_distance(query_vec, doc_vectors[i])
    print(f'Doc {i}: Cosine={cos_sim:6.4f}  |  Euclidean={euc_dist:6.4f}')
    print(f'  {documents[i]}')
    print()

Cosine Similarity vs Euclidean Distance
Query: "I love visiting tropical beaches in summer."

Doc 0: Cosine=1.0000  |  Euclidean=0.0000
  I love visiting tropical beaches in summer.

Doc 1: Cosine=0.6921  |  Euclidean=0.7848
  Mountain hiking is my favorite outdoor activity.

Doc 2: Cosine=0.7155  |  Euclidean=0.7543
  A sunny beach vacation is perfect for relaxation.

Doc 3: Cosine=0.6244  |  Euclidean=0.8667
  Rock climbing is an exciting sport.

Doc 4: Cosine=0.6869  |  Euclidean=0.7914
  Swimming in the ocean feels refreshing.

Doc 5: Cosine=0.6190  |  Euclidean=0.8729
  Winter skiing in the Alps is incredible.



## Metric Comparison Table

| Metric | Range | Use Case | Speed | Interpretation |
|--------|-------|----------|-------|-----------------|
| **Cosine Similarity** | [-1, 1] | Text/embeddings | Very Fast | Angle between vectors |
| **Euclidean Distance** | [0, ∞] | Image/general | Fast | Actual distance |
| **Dot Product** | [-∞, ∞] | FAISS indices | Fastest | Raw similarity (needs normalization) |
| **Manhattan Distance** | [0, ∞] | Sparse data | Very Fast | Sum of absolute differences |

### Best Practice
For text embeddings from language models: **Use cosine similarity**. It's the industry standard and works best with normalized embeddings.

In [6]:
# Build a similarity matrix
n_docs = len(documents)
similarity_matrix = np.zeros((n_docs, n_docs))

for i in range(n_docs):
    for j in range(n_docs):
        similarity_matrix[i][j] = cosine_similarity(doc_vectors[i], doc_vectors[j])

# Create a nice display
import pandas as pd

df_sim = pd.DataFrame(
    similarity_matrix,
    index=[f'Doc{i}' for i in range(n_docs)],
    columns=[f'Doc{i}' for i in range(n_docs)]
)

print('Similarity Matrix (Cosine):')
print(df_sim.round(3))
print()

# Find most and least similar pairs
max_sim = -2
min_sim = 2
max_pair = None
min_pair = None

for i in range(n_docs):
    for j in range(i+1, n_docs):
        if similarity_matrix[i][j] > max_sim:
            max_sim = similarity_matrix[i][j]
            max_pair = (i, j)
        if similarity_matrix[i][j] < min_sim:
            min_sim = similarity_matrix[i][j]
            min_pair = (i, j)

print('MOST SIMILAR PAIR:')
i, j = max_pair
print(f'  Doc {i}: {documents[i]}')
print(f'  Doc {j}: {documents[j]}')
print(f'  Similarity: {max_sim:.4f}')
print()
print('LEAST SIMILAR PAIR:')
i, j = min_pair
print(f'  Doc {i}: {documents[i]}')
print(f'  Doc {j}: {documents[j]}')
print(f'  Similarity: {min_sim:.4f}')

Similarity Matrix (Cosine):
       Doc0   Doc1   Doc2   Doc3   Doc4   Doc5
Doc0  1.000  0.692  0.716  0.624  0.687  0.619
Doc1  0.692  1.000  0.622  0.698  0.645  0.679
Doc2  0.716  0.622  1.000  0.594  0.687  0.605
Doc3  0.624  0.698  0.594  1.000  0.642  0.655
Doc4  0.687  0.645  0.687  0.642  1.000  0.614
Doc5  0.619  0.679  0.605  0.655  0.614  1.000

MOST SIMILAR PAIR:
  Doc 0: I love visiting tropical beaches in summer.
  Doc 2: A sunny beach vacation is perfect for relaxation.
  Similarity: 0.7155

LEAST SIMILAR PAIR:
  Doc 2: A sunny beach vacation is perfect for relaxation.
  Doc 3: Rock climbing is an exciting sport.
  Similarity: 0.5937


## Cost Comparison: Embedding Models (2026 Pricing)

| Model | Provider | Cost | Dimensions | Use Case |
|-------|----------|------|------------|----------|
| **gemini-embedding-001** | Google | $0.15 / 1M tokens | 768 | Good all-around, free tier available |
| **text-embedding-3-small** | OpenAI | $0.02 / 1M tokens | 1536 | Budget-friendly, good quality |
| **text-embedding-3-large** | OpenAI | $0.13 / 1M tokens | 3072 | Best quality, slower |
| **mistral-embed** | Mistral | $0.01 / 1M tokens | 1024 | Cheapest, good performance |
| **Local (all-MiniLM-L6-v2)** | Open source | Free | 384 | Private, requires GPU/CPU |

### Cost Example
Embedding 1 million documents (100 tokens each) = 100M tokens:
- **Google Gemini:** $15
- **OpenAI Small:** $2
- **Mistral:** $1
- **Local:** $0 (your compute)

### Selection Guide
- **Best quality:** OpenAI text-embedding-3-large
- **Best value:** Mistral Embed
- **Easiest setup:** Google Gemini (try free tier first)
- **Privacy required:** Local embeddings (all-MiniLM-L6)
- **Production RAG:** Mix of cheap (Mistral) + occasional re-ranking (OpenAI large)

## FAISS: Fast Approximate Nearest Neighbor Search

**FAISS** (Facebook AI Similarity Search) is an open-source library for fast similarity search.

### Problem It Solves
With 1 million documents, comparing query against every single embedding is *slow*.

```
Naive search: O(n*d)  where n=docs, d=dimensions
FAISS search:  O(log n) or O(√n)  ← MUCH faster!
```

### FAISS Benefits
- ✓ **Free & open source**
- ✓ **Runs locally** (no API calls)
- ✓ **Extremely fast** (even with millions of vectors)
- ✓ **Multiple index types** (exact, approximate, GPU)
- ✓ **Easy to use**

### Index Types
- **IndexFlatL2:** Exact search (brute force) - good for small datasets
- **IndexIVFFlat:** Approximate - partitions vectors into buckets
- **IndexHNSW:** Hierarchical navigation - best quality/speed tradeoff
- **GPU indices:** For very large scale (billions of vectors)

In [7]:
import faiss

# Prepare vectors (FAISS requires float32)
dimension = len(doc_vectors[0])  # Should be 768
vectors = np.array(doc_vectors).astype('float32')

print(f'Creating FAISS index...')
print(f'  Vectors shape: {vectors.shape}')
print(f'  Dimension: {dimension}')
print()

# Create a simple flat index (exact nearest neighbors)
index = faiss.IndexFlatL2(dimension)
index.add(vectors)

print(f'✓ Index built successfully!')
print(f'  Total vectors in index: {index.ntotal}')

ModuleNotFoundError: No module named 'faiss'

## Searching with FAISS

Now we'll use FAISS to find the most similar documents to a query.

In [ ]:
# Create a query vector
query_text = 'I want to relax on a beach'
query_vector = embeddings.embed_query(query_text)
query_vector = np.array([query_vector]).astype('float32')

print(f'Query: "{query_text}"')
print()

# Search for k=3 nearest neighbors
k = 3
distances, indices = index.search(query_vector, k)

print(f'Top {k} results:')
print('=' * 70)
for rank, (idx, dist) in enumerate(zip(indices[0], distances[0])):
    idx = int(idx)
    print(f'{rank + 1}. Distance: {dist:.4f}')
    print(f'   Document: {documents[idx]}')
    # Convert L2 distance back to similarity for interpretation
    cos_sim = cosine_similarity(query_vector[0], doc_vectors[idx])
    print(f'   (Cosine sim: {cos_sim:.4f})')
    print()

## Using FAISS with LangChain (Easier!)

LangChain provides a clean wrapper around FAISS, making integration with embeddings seamless.

In [ ]:
from langchain_community.vectorstores import FAISS

print('Creating LangChain FAISS vectorstore...')
print()

# Create vectorstore from documents
vectorstore = FAISS.from_texts(documents, embeddings)

print(f'✓ Vectorstore created with {len(documents)} documents')
print()

# Search with a natural language query
query = 'Beach vacation'
results = vectorstore.similarity_search(query, k=3)

print(f'Search results for: "{query}"')
print('=' * 70)
for i, doc in enumerate(results):
    print(f'{i + 1}. {doc.page_content}')
    print()

## Building a Simple Document Search Engine

Let's create a complete, end-to-end semantic search system with 10 travel-related documents.

In [8]:
# Our knowledge base: travel advice documents
knowledge_base = [
    'Paris is famous for the Eiffel Tower, museums, and world-class restaurants.',
    'Tokyo offers neon-lit streets, sushi bars, and ancient temples nearby.',
    'Bali has stunning beaches, rice terraces, and affordable accommodations.',
    'Iceland features dramatic waterfalls, glaciers, and Northern Lights.',
    'Thailand is known for tropical islands, temples, and spicy street food.',
    'Switzerland offers skiing in winter and hiking in summer with alpine scenery.',
    'Egypt is home to the Pyramids, Cairo museums, and Nile River cruises.',
    'Australia has Great Barrier Reef, desert outback, and unique wildlife.',
    'Norway offers fjords, Northern Lights, and outdoor adventure activities.',
    'Indonesia has volcanoes, temples, and diverse island cultures.',
]

print(f'Knowledge base loaded: {len(knowledge_base)} documents')
for i, doc in enumerate(knowledge_base):
    print(f'{i+1:2}. {doc[:60]}...')

Knowledge base loaded: 10 documents
 1. Paris is famous for the Eiffel Tower, museums, and world-cla...
 2. Tokyo offers neon-lit streets, sushi bars, and ancient templ...
 3. Bali has stunning beaches, rice terraces, and affordable acc...
 4. Iceland features dramatic waterfalls, glaciers, and Northern...
 5. Thailand is known for tropical islands, temples, and spicy s...
 6. Switzerland offers skiing in winter and hiking in summer wit...
 7. Egypt is home to the Pyramids, Cairo museums, and Nile River...
 8. Australia has Great Barrier Reef, desert outback, and unique...
 9. Norway offers fjords, Northern Lights, and outdoor adventure...
10. Indonesia has volcanoes, temples, and diverse island culture...


In [9]:
# Embed all knowledge base documents
print('Embedding documents...')
kb_vectors = [embeddings.embed_query(doc) for doc in knowledge_base]
kb_vectors = np.array(kb_vectors).astype('float32')

# Build FAISS index
kb_index = faiss.IndexFlatL2(kb_vectors.shape[1])
kb_index.add(kb_vectors)

print(f'✓ Index ready!')
print(f'  {kb_index.ntotal} documents indexed')
print(f'  {kb_vectors.shape[1]} dimensions')

Embedding documents...


NameError: name 'faiss' is not defined

In [ ]:
def search_documents(query, index, documents, k=3):
    """Search the knowledge base for similar documents."""
    # Embed query
    query_vec = embeddings.embed_query(query)
    query_vec = np.array([query_vec]).astype('float32')
    
    # Search FAISS index
    distances, indices = index.search(query_vec, k)
    
    # Format results
    results = []
    for idx, dist in zip(indices[0], distances[0]):
        results.append({
            'rank': len(results) + 1,
            'document': documents[int(idx)],
            'distance': float(dist)
        })
    return results

# Test the search function
test_queries = [
    'Best beaches in tropical places',
    'Northern lights and winter activities',
    'Ancient historical monuments'
]

for query in test_queries:
    print(f'Query: "{query}"')
    print('=' * 70)
    results = search_documents(query, kb_index, knowledge_base, k=3)
    for res in results:
        print(f'{res["rank"]}. {res["document"]}')
    print()

## Saving and Loading the Index

To avoid re-embedding documents every time, save the FAISS index to disk.

In [ ]:
# Save the index
index_path = '/tmp/travel_search_index.faiss'
faiss.write_index(kb_index, index_path)
print(f'✓ Index saved to {index_path}')
print()

# Load the index
loaded_index = faiss.read_index(index_path)
print(f'✓ Index loaded successfully')
print(f'  Contains {loaded_index.ntotal} vectors')
print()

# Verify it works
test_query = 'tropical paradise'
results = search_documents(test_query, loaded_index, knowledge_base, k=2)
print(f'Test query: "{test_query}"')
for res in results:
    print(f'  {res["rank"]}. {res["document"]}')

## Exercise 1: FAQ Search System

**Goal:** Build a simple customer support FAQ search engine.

**Instructions:**
1. Create a list of FAQ questions and answers
2. Embed the questions
3. When a customer asks a question, find the closest FAQ match
4. Return the answer

**Example:**
- Customer: "How do I return an item?"
- System finds: "What is your return policy?" (FAQ #3)
- System returns: "Items can be returned within 30 days..."

**Tips:**
- Use FAISS IndexFlatL2 for small FAQ sets (<1000)
- Cosine similarity is industry standard
- Add confidence threshold (if similarity too low, say "I don't know")

In [ ]:
# Exercise 1: FAQ Search System
# TODO: Implement FAQ search

# Step 1: Create FAQ dataset
faqs = [
    # Q: "...", A: "..."
    # Add your FAQs here
]

# Step 2: Embed FAQ questions
# faq_vectors = ...

# Step 3: Build FAISS index
# faq_index = ...

# Step 4: Create search function
# def find_faq_answer(user_question):
#     ...

print('TODO: Complete the FAQ search implementation')

## Exercise 2: Compare Similarity Metrics

**Goal:** Empirically compare cosine similarity vs Euclidean distance.

**Instructions:**
1. Create 5-10 document pairs with varying similarity
2. For each pair, compute:
   - Cosine similarity
   - Euclidean distance
   - Normalized Euclidean distance
3. Plot the results (or create a table)
4. Discuss when each metric is best

**Hypothesis:**
For normalized embeddings, cosine and euclidean should rank documents similarly.

In [ ]:
# Exercise 2: Compare Similarity Metrics
# TODO: Implement metric comparison

# Step 1: Create document pairs
comparison_docs = [
    # 'doc text 1',
    # 'doc text 2',
]

# Step 2: Embed documents
# comparison_vectors = ...

# Step 3: Compare metrics
# for pair in ...
#     cosine = ...
#     euclidean = ...
#     normalized_euclidean = ...

print('TODO: Complete the metric comparison')

## Session Summary: Vector Search 101

### Key Takeaways
1. **Embeddings** convert text to numerical vectors that capture semantic meaning
2. **Cosine similarity** is the industry standard for comparing embeddings
3. **FAISS** enables fast nearest-neighbor search on large datasets
4. **Vector search** powers modern RAG systems, recommendation engines, and semantic search

### What You've Learned
- ✓ Creating embeddings with Google Gemini
- ✓ Computing similarity metrics (cosine, Euclidean, dot product)
- ✓ Building and searching FAISS indices
- ✓ Integrating with LangChain for production use
- ✓ Cost comparison of embedding models

### Next Steps (Session 11: RAG Fundamentals)
- Combine vector search with LLMs for retrieval-augmented generation
- Build a document Q&A system
- Handle document chunking and metadata
- Implement reranking for better quality

### Resources
- [FAISS Documentation](https://github.com/facebookresearch/faiss)
- [LangChain Vector Stores](https://python.langchain.com/docs/modules/data_connection/vectorstores/)
- [Embedding Models Benchmark](https://huggingface.co/spaces/mteb/leaderboard)
- [Vector Database Guide](https://www.pinecone.io/learn/vector-database/)